### ISSUE TITLE
Problem changing locale with multiple threads

### ISSUE THREAD

**Issue by michelboaventura** [svenfuchs/i18n#269]
Hi,

I'm having a problem with I18n and multiple threads on a pure ruby app, which can be reproduced with this small script:

``` ruby
require 'i18n'

I18n.available_locales = [:pt, :en]
I18n.enforce_available_locales = true

t1 = Thread.new do
  1.upto(4) do
    puts I18n.t(:foo)
    sleep 1
  end
end

sleep 1 #<<<THIS

I18n.default_locale = :pt
I18n.locale = :pt

t1.join
```

The problem is: If I run this, I got as output four `translation missing: en.foo`, but if I remove the `sleep` marked on the code, I got four `translation missing: pt.foo`.

I'm aware that the four messages for the pt make sense, since the locale change happens before the thread runs, but I think that on the first case it should print some messages on `en` and some on `pt` (the amount of each would be not deterministic).

This is a expected behavior?

**Comment by boazsegev** [comment]
The issue is still unresolved.

The source of the issue, as it seems from my testing, is that I18n stores the settings in the thread's variable using: `Thread.current[:i18n_config]`.

As you can see in [the source code](https://github.com/mattetti/i18n/blob/32eed2028c1d6148fd0de2079ded40da740ae259/lib/i18n.rb#L18), the settings are persistent. The default_locale isn't reviewed again after it had been reviewed for the first time.

The solution would probably be to change the [I18n::Config class's #locale method](https://github.com/mattetti/i18n/blob/32eed2028c1d6148fd0de2079ded40da740ae259/lib/i18n/config.rb#L6) to keep accessing the `default_locale` (replacing the `||=` with `||`), like so:

```ruby
@locale || default_locale
```

An easy workaround is to manually reset the locale to the default locale each time.

Here is the code I used for testing:

```ruby
require 'i18n'

I18n.available_locales = [:pt, :en]
I18n.enforce_available_locales = true
I18n.default_locale = :en

# t1 = Thread.new do
#   1.upto(4) do
#     Thread.current[:i18n_config] = nil
#     puts I18n.t(:foo)
#     puts Thread.current.keys
#     sleep 1
#   end
# end

t1 = Thread.new do
  1.upto(4) do
    I18n.locale = I18n.default_locale
    puts I18n.t(:foo)
    sleep 1
  end
end

sleep 0.1 #<<<THIS

I18n.default_locale = :pt

t1.join
```

**Comment by radar** [comment]
I've added the fix from #320 to master now, which should fix this issue. Thanks everyone!
